### PR TITLE
render exceptions in conditions as condition failure (copy of #49)

### PR DESCRIPTION
**Pull request by Fuud**
This pull request is copy of #49 but rebased on master.
Please post any questions/feedbacks to #49 to keep all in one place.

### CODE REVIEW

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java [hunk 1]
```diff
@@ -530,15 +583,53 @@ private Expression rewriteStaticMethodCondition(StaticMethodCallExpression condi
...
+
+    tryCatchStatement.addCatch(
+        new CatchStatement(
+            new Parameter(new ClassNode(Throwable.class), "throwable"),
+            new ExpressionStatement(
+                AstUtil.createDirectMethodCall(
+                    new ClassExpression(resources.getAstNodeCache().SpockRuntime),
+                    resources.getAstNodeCache().SpockRuntime_ConditionFailedWithException,
+                    new ArgumentListExpression(Arrays.asList(
+                        message == null ? new VariableExpression("$spock_valueRecorder") : ConstantExpression.NULL, // recorder
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: is there a reason why "$spock_valueRecorder" cannot be a constant, if not then it should be refactored.

File: spock-core/src/main/java/org/spockframework/runtime/ValueRecorder.java [hunk 2]
```diff
@@ -27,6 +27,7 @@
  */
 public class ValueRecorder {
   private final ArrayList<Object> values = new ArrayList<Object>();
+  private final Stack<Integer> startedRecordings = new Stack<Integer>();
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/ValueRecorder.java]: Why use `Stack` here? I should be `Deque<Integer> startedRecordings= new ArrayDeque<Integer>();`

File: spock-specs/src/test/groovy/org/spockframework/smoke/condition/ExceptionsInConditions.groovy [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                  def map = new HashMap<String, String>()
+                  expect:
+                  getKey(map, "key").length() == 0
+                """)
+
+    then:
+    SpockException e = thrown()
+    e.getMessage() == "key does not exists"
+  }
+}
```
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/smoke/condition/ExceptionsInConditions.groovy]: Could you add a test with deeper nesting, `a.b.c.d.e()` where b is null.

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java [hunk 4]
```diff
@@ -486,13 +495,33 @@ private Expression unrecord(Expression expr) {
...
+  // extractVariableNumber(record(expr)) ==
+  // extractVariableNumber($spock_valueRecorder.record($spock_valueRecorder.startRecordingValue(variableNumber), expr) ==
+  // variableNumber
+  private int extractVariableNumber(Expression expr){
+    if (!(expr instanceof MethodCallExpression)) return -1;
+    MethodCallExpression methodExpr = (MethodCallExpression) expr;
+    Expression targetExpr = methodExpr.getObjectExpression();
+    if (!(targetExpr instanceof VariableExpression)) return -1;
+    VariableExpression var = (VariableExpression)targetExpr;
+    if (!var.getName().equals("$spock_valueRecorder")) return -1;
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: equality on strings should always be "constant".equals(variable)
`!"$spock_valueRecorder".equals(var.getName())`

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java [hunk 5]
```diff
@@ -566,4 +657,22 @@ private Expression rewriteToSpockRuntimeCall(MethodNode method, Expression condi
     result.setSourcePosition(condition);
     return result;
   }
+
+  private class Unrecorded{
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: Where is this class used?

File: spock-core/src/main/java/org/spockframework/runtime/SpockException.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+  public SpockException(String message, Throwable cause) {
+    super(message, cause);
+  }
+
+  public SpockException(Throwable cause) {
+    super(cause);
+  }
+
+//  protected SpockException(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace) {
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/SpockException.java]: Commented out code

File: spock-core/src/main/java/org/spockframework/runtime/ExpressionInfoBuilder.java [hunk 7]
```diff
@@ -17,6 +17,7 @@
 package org.spockframework.runtime;
 
 import java.util.Iterator;
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/ExpressionInfoBuilder.java]: Now unused

File: spock-core/src/main/java/org/spockframework/runtime/ExpressionInfoBuilder.java [hunk 8]
```diff
@@ -34,13 +36,17 @@
   private final String text;
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/ExpressionInfoBuilder.java]: Private field 'text' is assigned but never accessed

File: spock-core/src/main/java/org/spockframework/runtime/JUnitSupervisor.java [hunk 9]
```diff
@@ -116,16 +116,19 @@ private int handleMultipleFailures(ErrorInfo error) {
   private boolean isFailedEqualityComparison(Throwable exception) {
     if (!(exception instanceof ConditionNotSatisfiedError)) return false;
 
-    Condition condition = ((ConditionNotSatisfiedError) exception).getCondition();
+    final ConditionNotSatisfiedError conditionNotSatisfiedError = (ConditionNotSatisfiedError) exception;
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/JUnitSupervisor.java]: Why is this final and condition not? Same issue on line 130

File: spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java [hunk 10]
```diff
@@ -34,21 +34,44 @@ public static void verifyCondition(@Nullable ValueRecorder recorder,
...
+  public static final String CONDITION_FAILED_WITH_EXCEPTION = "conditionFailedWithException";
+
+  public static void conditionFailedWithException(@Nullable ValueRecorder recorder, @Nullable String text, int line, int column, @Nullable Object message, Throwable throwable){
+    if (throwable instanceof SpockAssertionError){
+      throw (SpockAssertionError) throwable; // this is our exception - it already has good message
+    }
+    if (throwable instanceof SpockException){
+      throw (SpockException) throwable; // this is our exception - it already has good message
+    }
+    throw new ConditionNotSatisfiedError(
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java]: What about JVM errors (OutOfMemoryError)? They will be converted to harmless RuntimeExceptions.

File: spock-core/src/main/java/org/spockframework/runtime/ValueRecorder.java [hunk 11]
```diff
@@ -44,9 +45,29 @@ public ValueRecorder reset() {
...
+      }
+    }
+    if (!foundThisCallOnStack){
+      throw new IllegalStateException("can not found call #"+index+" on stack. Invalid call of record method?");
+    }
+
     return value;
   }
 
+  public static final String START_RECORDING_VALUE = "startRecordingValue";
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/ValueRecorder.java]: Isn't it more like `startRecordingIndex`?

**Comment by leonard84** [comment]
There are many formatting issues, but since we want to fix the whole project with #555 I'd be ok with leaving them for now.

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java (hunk 1 above)
Comment (Fuud) [inline spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: It can. Let it be separate PR.

File: spock-core/src/main/java/org/spockframework/runtime/ValueRecorder.java (hunk 2 above)
Comment (Fuud) [inline spock-core/src/main/java/org/spockframework/runtime/ValueRecorder.java]: Why not? We need stack here with pop and push operation.

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java (hunk 4 above)
Comment (Fuud) [inline spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: May be. Should I change all such places in this file?

File: spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java (hunk 10 above)
Comment (Fuud) [inline spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java]: They will be converted to ConditionNotSatisfiedError. It is good for tests.

File: spock-core/src/main/java/org/spockframework/runtime/ValueRecorder.java (hunk 11 above)
Comment (Fuud) [inline spock-core/src/main/java/org/spockframework/runtime/ValueRecorder.java]: No. Spock start record value in this point.

File: spock-core/src/main/java/org/spockframework/runtime/ValueRecorder.java (hunk 2 above)
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/ValueRecorder.java]: http://docs.oracle.com/javase/8/docs/api/java/util/Stack.html
> A more complete and consistent set of LIFO stack operations is provided by the Deque interface and its implementations, which should be used in preference to this class.

And Stack is based on Vector:
> Unlike the new collection implementations, Vector is synchronized. If a thread-safe implementation is not needed, it is recommended to use ArrayList in place of Vector.

File: spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java (hunk 10 above)
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java]: No if the JVM throws OOME there is a reason for that, any operation after that may fail, e.g., rendering the output. See https://stackoverflow.com/questions/8728866/no-throw-virtualmachineerror-guarantees/8729011#8729011

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java (hunk 1 above)
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: OK

File: spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java (hunk 10 above)
Comment (Fuud) [inline spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java]: Yes and No.
It dependends why OOM occured. If it based on method parameters it will be good to provide parameters values to user thus he will be able to determine what is going wrong. Same for example for StackOverflowError. If OOM occures independently from test code than there is nothing we can do and does not matter when and where OOM occure again.

File: spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java (hunk 4 above)
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java]: yes

File: spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java (hunk 10 above)
Comment (szpak) [inline spock-core/src/main/java/org/spockframework/runtime/SpockRuntime.java]: I'm not an expert in a topic of OOM, so I may be wrong, however it some cases it could be not necessary to shutdown the whole JVM in case of OOM (like a try to allocate a large memory block in a particular thread). In that case a try to render it as a "normal" error could succeed. Even if not another OOM error will be thrown (maybe with less interesting stack trace), but it is just a test - it should be possible to find out why it happened anyway. There is a risk to keep tests execution in an undefined state, but I think we talk here about quite rare situations and the more programmatic approach would to keep this useful PR as is to make it easier/faster to merge and - if needed - come back to that topic after the first related bug report. What do you think @leonard84?
